src/shared_memory_manager.cc

```diff
@@ -60,72 +238,202 @@ SharedMemoryManager::GetCUDAHandle(
+    }
+    RETURN_IF_ERR(shm_region.AddUInt("byte_size", it->second->byte_size_));
+    RETURN_IF_ERR(shm_status->Append(std::move(shm_region)));
+  }
+
+  return nullptr;  // success
 }
 
 TRITONSERVER_Error*
 SharedMemoryManager::Unregister(
     const std::string& name, TRITONSERVER_MemoryType memory_type)
 {
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
-}
+  // Serialize all operations that write/read current shared memory regions
+  std::lock_guard<std::mutex> lock(mu_);
 
-TRITONSERVER_Error*
-SharedMemoryManager::UnregisterAll(TRITONSERVER_MemoryType memory_type)
-{
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
+  return UnregisterHelper(name, memory_type);
 }
 
 TRITONSERVER_Error*
 SharedMemoryManager::UnregisterHelper(
     const std::string& name, TRITONSERVER_MemoryType memory_type)
 {
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
+  // Must hold the lock on register_mu_ while calling this function.
+  auto it = shared_memory_map_.find(name);
+  if (it != shared_memory_map_.end() && it->second->kind_ == memory_type) {
+    if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+      RETURN_IF_ERR(UnmapSharedMemory(it->second->mapped_addr_));
+    } else {
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_INVALID_ARG,
+          std::string(
+              "failed to unregister CUDA shared memory region: '" + name +
+              "', GPUs not supported")
+              .c_str());
+    }
+
+    // Remove region information from shared_memory_map_
+    shared_memory_map_.erase(it);
+  }
+
+  return nullptr;
 }
-}}  // namespace triton::server
-#else
-#include <errno.h>
-#include <fcntl.h>
-#include <sys/mman.h>
-#include <unistd.h>
 
-#include "common.h"
-#include "triton/common/logging.h"
+TRITONSERVER_Error*
+SharedMemoryManager::UnregisterAll(TRITONSERVER_MemoryType memory_type)
+{
+  std::lock_guard<std::mutex> lock(mu_);
+  std::string error_message = "Failed to unregister the following ";
+  std::vector<std::string> unregister_fails;
+  if (memory_type == TRITONSERVER_MEMORY_CPU) {
+    // Serialize all operations that write/read current shared memory regions
+    error_message += "system shared memory regions: ";
+    for (auto it = shared_memory_map_.cbegin(), next_it = it;
+         it != shared_memory_map_.cend(); it = next_it) {
+      ++next_it;
+      if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+        TRITONSERVER_Error* err = UnregisterHelper(it->first, memory_type);
+        if (err != nullptr) {
+          unregister_fails.push_back(it->first);
+        }
+      }
+    }
+  }
```

Comment:
Add a comment about checking GPU shared memory here and to check if we still need a windows implementation of `UnregisterAll` once we have GPU shared memory support

Reply:
Apologies, I was concerned with the other changes and am looking at this comment post-push. Do you recall where this comment was in reference to?

Reply:
This is in reference to the if-statement checking if the memory type is CPU. At the end of the if-block, where the `else` statement would go, I would like a comment stating that we will check for the GPU memory type once it's available. Also, I would like a comment to revisit this methods being a duplicate of the linux implementation once we check for the GPU type.